include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,162 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+void thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event);
+void thread_event_update(tsd_t *tsd);
+
+/*
+ * List of all event wait times, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+/*
+ * List of all thread event counters, in the following format:
+ *  C(counter,				tsd_field)
+ */
+#define W(event, condition)						\
+    C(event##_event_wait,		event##_event_wait)
+#define ITERATE_OVER_ALL_COUNTERS					\
+    C(thread_allocated,			thread_allocated)		\
+    C(thread_allocated_threshold_fast,	thread_allocated_threshold_fast)\
+    C(thread_allocated_last_event,	thread_allocated_last_event)	\
+    C(thread_allocated_next_event,	thread_allocated_threshold)	\
```

Comment:
You mean renaming the original `tsd` fields and not renaming them here, or keeping the original `tsd` fields and renaming them here?

Also what can be a good name for `thread_allocated_threshold_fast`? `thread_allocated_next_event_fast`?

Reply:
I mean better to make it consistent, i.e. using next_event instead of threshold in tsd as well.

thread_allocated_next_event_fast should be fine. It’s only referenced in a handful of places, so the length don’t bother me.